contrib-compiler.md

```diff
@@ -0,0 +1,99 @@
+
+
+Thanks for interest in contributions.
+
+Note currently our philosphy is to make release build as easy as possible, while dev build is fairly complex at this time.
+
+## Set up dev dependencies
+
+- Having [https://opam.ocaml.org/](opam) installed
+
+    opam switch 4.02.3+buckle-master # use our OCaml compiler
+    opam install camlp4 cppo  <1>
+
+   Both `Camlp4` and `cppo` are used to generate OCaml code for pre-processing. 
+   Switch to our compiler makes your life easier (avoid confusion with diferent compiler versions)
+- Having [https://nodejs.org/](NodeJS) installed
+- Having GNU Make installed
+- OS: Mac/Linux (Note BuckleScript works well on Windows, but the dev mode is not tested)
+- Our dev build depends on a submodule
+
+  ```sh
+  git submodule --update
+  ``` 
+
+All source code are in `jscomp` directory, below assume that you are working in `jscomp` directory.
+
+
+## components of the compiler
+
+`bs-platform` is composed of several components, the core compiler, the bundled ocaml stdlib,  shipped js bindings, test cases and some internal tools.
+
+### internal tools
+
+- bspack
+
+  bspack is heavily used in `bs-platform`, it is a tool to pack all ocaml 
+  source code  into a single file, so that in release mode, there is not any dev dependency anymore.
+
+  It is very stable and heavily tested, in most cases, contributors don't need change its behavior.
+
+  Below are examples of how it is used:
+
+  ```make
+  bin/whole_compiler.ml:./bin/bspack.exe
+	$< -U BS_DEBUG -bs-MD -prelude-str 'module Config = Config_whole_compiler' -bs-exclude-I config -o $@ -bs-main Js_main -I ../ocaml/utils/ -I ../ocaml/parsing/ -I ../ocaml/typing/ -I ../ocaml/bytecomp/ -I ../ocaml/driver/ -I stubs -I ext -I syntax -I depends -I common -I core
+  ```
+
+  `bspack` is bootstrapped by itself, so it is only a single file
+  to build `bspack.exe`, run 
+  
+  ```sh
+  make -C bin bspack.exe
+  ```
+## Dev-Build the compiler binaries
+`bspack` makes release build much easier, it is pretty slow during development since it does not do separate compilation. For dev build, we build each component and link them together.
+
+Please refer `./build.sh` for more details.
+
+Inside `build.sh`, it streamlined the whole process,
+first, it setup env vars, and checks all files type check:
+
+```sh
+make check
+```
+
+Then it link the compiler and build system
+```sh
+make bin/bsc.exe bin/bsb.exe
+```
+
+Note `make bin/bsc.exe` is different from `make -C bin bsc.exe`, the former is the dev build, the latter is the release build.
+
+Then it starts building libs
+```sh
+make libs
+```
+
+Then it starts building tests
+```
+make -C test all
+```
+
+After all these finished, it start snapshoting the changes using bspack, updating the dependency (prepare for the next build):
+
+```
+make depend snapshotml
+cd runtime; make depend; cd ..
+cd others; make depend; cd ..  
+```
+
```

Comment:
suggestion: A note here saying that the build log is written to `jscomp/build.compile` could be helpful for debugging, that's a nice feature of the build.